edgehog-device-runtime-forwarder/src/connection.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 SECO Mind Srl
+// SPDX-License-Identifier: Apache-2.0
+
+//! Manage a single connection.
+//!
+//! A connection is responsible for sending and receiving data through a WebSocket connection from
+//! and to the [`ConnectionsManager`](crate::connections_manager::ConnectionsManager).
+
+use std::ops::{Deref, DerefMut};
+
+use thiserror::Error as ThisError;
+use tokio::sync::mpsc::Sender;
+use tokio::task::{JoinError, JoinHandle};
+use tracing::{debug, error, instrument, span, Level};
+
+use crate::messages::{
+    Http as ProtoHttp, HttpMessage as ProtoHttpMessage, HttpResponse, Id, ProtoMessage,
+    ProtocolError,
+};
+
+/// Connection errors.
+#[non_exhaustive]
+#[derive(displaydoc::Display, ThisError, Debug)]
+pub enum ConnectionError {
+    /// Channel error.
+    ChannelToWs,
+    /// Reqwest error.
+    Reqwest(#[from] reqwest::Error),
+    /// Protobuf error.
+    Protobuf(#[from] ProtocolError),
+    /// Failed to Join a task handle.
+    JoinError(#[from] JoinError),
+}
+
+/// Handle to the task spawned to handle a [`Connection`].
+#[derive(Debug)]
+pub(crate) struct ConnectionHandle {
+    /// Handle of the task managing the connection.
+    pub(crate) handle: JoinHandle<()>,
+}
+
+impl Deref for ConnectionHandle {
+    type Target = JoinHandle<()>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.handle
+    }
+}
+
+impl DerefMut for ConnectionHandle {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.handle
+    }
+}
+
+/// Struct containing a connection information useful to communicate with the
+/// [`ConnectionsManager`](crate::collection::ConnectionsManager).
+#[derive(Debug)]
+pub(crate) struct Connection {
+    id: Id,
+    tx_ws: Sender<ProtoMessage>,
+    request: reqwest::RequestBuilder,
+}
+
+impl Connection {
+    /// Initialize a new connection.
+    pub(crate) fn new(
+        id: Id,
+        tx_ws: Sender<ProtoMessage>,
+        request: reqwest::RequestBuilder,
+    ) -> Self {
+        Self { id, tx_ws, request }
+    }
+
+    /// Spawn the task responsible for handling the connection.
+    #[instrument(skip_all)]
+    pub(crate) fn spawn(self) -> ConnectionHandle {
+        // spawn a task responsible for notifying when new data is available
+        let handle = tokio::spawn(async move {
+            // the span in used to know from which request a possible error is generated
+            let span = span!(Level::DEBUG, "spawn", id = %self.id);
+            let _enter = span.enter();
+
+            if let Err(err) = self.task().await {
+                error!("connection task failed with error {err:?}");
+            }
```

Comment:
This should be in a function with instrument, to be thread local compatible (impl detail for tokio and tracing)